specs/interop.md

```diff
@@ -0,0 +1,188 @@
+- User deposits from L1
+- Network Upgrade Transactions
+  - L1Block deployment
+  - Update L1Block Proxy ERC-1967 Implementation Slot
+
+The execution payload MUST set `noTxPool` to `false` for this block.
+
+The exact definitions for these upgrade transactions are still to be defined.
+
+## Derivation Pipeline
+
+The derivation pipeline enforces invariants on safe blocks that include relaying transactions.
+
+- The relaying message MUST have a corresponding initiating message
+- The initiating message that corresponds to a relaying message MUST come from an allowed chain
+
+Blocks that contain transactions that relay cross domain messages to the remote chain where the
+initiating transaction does not exist MUST be considered invalid and MUST not be allowed by the
+derivation pipeline to be considered safe.
+
+There is no concept of replay protection within the protocol since there is no replay protection
+mechanism that fits well for all applications. Users MAY submit an arbitrary number of relaying
+messages per initiating message. This greatly simplifies the protocol and allows for
+applications to build application specific replay protection.
+
+### Depositing a Relaying Message
+
+Deposit transactions (force inclusion transactions) give censorship resistance to layer two networks.
+The derivation pipeline must gracefully handle the case in which a user uses a deposit transaction to
+relay a cross chain message. To not couple preconfirmation security to consensus, deposit transactions
+that relay cross chain messages MUST have an initiating message that is considered safe. This relaxes
+a strict synchrony assumption on the sequencer that it MUST have all unsafe blocks of remote chains
+as fast as possible to ensure that it is building correct blocks. It also prevents a class of attacks
+where the user can send a relaying message before the initiating message and trick the derivation pipeline
+into reorganizing the sequencer.
+
+### Safety
+
+The initiating messages for all relaying messages MUST be resolved as safe before an L2 block can transition from being unsafe to safe.
+Users MAY optimistically accept unsafe blocks without any verification of the relaying messages. They SHOULD optimistically verify
+the initiating messages exist in remote unsafe blocks to more quickly reorganize out invalid blocks.
+
+## Predeploys
+
+Two new system level predeploys are introduced for managing cross chain messaging.
+
+### CrossL2Outbox
+
+Address: `0x4200000000000000000000000000000000000023`
+
+The `CrossL2Outbox` is responsible for initiating a cross chain message on the local chain.
+For every cross chain message that is sent, it MUST emit an event that indicates the existence
+of the cross chain message.
+
+An initiating message 
+
+### CrossL2Inbox
+
+Address: `0x4200000000000000000000000000000000000022`
+
+The `CrossL2Inbox` is responsible for relaying a cross chain message on the remote chain.
```

Comment:
@protolambda @tynes

Carrying over the principle that we should be creating a design that works for us first (Messaging & StandardBridge).

In slack, @karlfloersch mentioned leaving the state bloat problem as a future improvement is fine. Since the primary use case of interop is the bridge and this use case requires replay protection, why not simply have the `CrossL2Inbox` record these message hashes such that we can make the `CrossDomainMessenger` stateless. My reasoning for this

1. We have single expected op-chain behavior for deposits, withdrawals, interop transactions. These transactions are processed only once. It feels like strange to have a fork in protocol UX for interop transcations
2. It becomes a nice improvement to the CrossDomainMessenger and simplifies it's implementation (less code is better)

Reply:
All applications that want the same sort of replay protection that we want as the developers of the bridge can use the messenger. The messenger is responsible for replay protection and can give it to any application that needs replay protection. For people that want to use the lower level interface, they will need to implement their own type of replay protection. The idea is that the second order effect of forcing everybody to use our replay protection scheme is that it will end up bloating the state more

Reply:
> The idea is that the second order effect of forcing everybody to use our replay protection scheme is that it will end up bloating the state more

This is only true if there are even any apps that would use a lower-level interface without replay protection. Is there any supported evidence of apps like this

This also leads into another point that the state bloat still exists at all all message receipts from all remotes must be permanently retained since they can be relayed as many times as they want (even for msgs that revert) as the revert is at the evm level

Reply:
> This is only true if there are even any apps that would use a lower-level interface without replay protection. Is there any supporting evidence of apps like this that has any substantial usage?

This is also true if the particular type of replay protection that we use is useful for all applications. Which is fundamentally not true. See [this](https://github.com/ethereum-optimism/specs/pull/6#discussion_r1468216950) comment

Message receipts do not lead to state bloat as they are not part of state. They are part of history. It is a well known issue that history grows forever, which is why eip4444 exists. We will need to contribute to building eip4444 eventually

Reply:
forsure, even with eip4444 since a message can be replayed >1. Verifiers must have _some_ remote source of pruned history if someone choose to relay a really old message. Just wanted to note this concern

thanks for the distinction between state/history